--- original/gml_GlobalScript_scrGetRecords.gml
+++ patched/gml_GlobalScript_scrGetRecords.gml
@@ -35,6 +35,11 @@
             _records[i] = scrString("bar_no_record");
         }
         _records[i] = string_copy(_records[i], 1, 16);
+    }
+    with (obj_richpresence)
+    {
+        array_copy(records, 0, _records, 0, array_length(_records));
+        alarm[0] = 1;
     }
     return _records;
 }
--- original/gml_GlobalScript_scrPlayGame.gml
+++ patched/gml_GlobalScript_scrPlayGame.gml
@@ -10,6 +10,10 @@
     {
         return false;
     }
+    with (obj_richpresence)
+    {
+        alarm[0] = 1;
+    }
     global.currGame = arg0;
     global.currGameID = global.mGameID[arg0];
     global.solutionState = false;
--- original/gml_GlobalScript_scrSaveGame.gml
+++ patched/gml_GlobalScript_scrSaveGame.gml
@@ -36,6 +36,7 @@
         }
     }
     var _makeBackup = false;
+    scrGetRecords();
     if (global.backupTimer >= global.BACKUP_MINIMUM_TIME && global.currFile >= 1 && global.currFile <= global.NUM_PROFILES_ACCESSIBLE)
     {
         global.backupSaveNum[global.currFile]++;
