--- original/gml_GlobalScript_scrInitInput.gml
+++ patched/gml_GlobalScript_scrInitInput.gml
@@ -48,7 +48,63 @@
     global.DEFAULT_JOY[1][global.INP_FIRE1] = 32771;
     global.DEFAULT_JOY[1][global.INP_FIRE2] = 32769;
     global.DEFAULT_JOY[1][global.INP_START] = 32778;
-    for (p = 0; p <= 1; p++)
+    global.DEFAULT_KEY[2][global.INP_UP] = 73;
+    global.DEFAULT_KEY[2][global.INP_DOWN] = 75;
+    global.DEFAULT_KEY[2][global.INP_LEFT] = 74;
+    global.DEFAULT_KEY[2][global.INP_RIGHT] = 76;
+    global.DEFAULT_KEY[2][global.INP_FIRE1] = 85;
+    global.DEFAULT_KEY[2][global.INP_FIRE2] = 73;
+    global.DEFAULT_KEY[2][global.INP_START] = 33;
+    global.DEFAULT_JOY[2][global.INP_UP] = 32781;
+    global.DEFAULT_JOY[2][global.INP_DOWN] = 32782;
+    global.DEFAULT_JOY[2][global.INP_LEFT] = 32783;
+    global.DEFAULT_JOY[2][global.INP_RIGHT] = 32784;
+    global.DEFAULT_JOY[2][global.INP_FIRE1] = 32771;
+    global.DEFAULT_JOY[2][global.INP_FIRE2] = 32769;
+    global.DEFAULT_JOY[2][global.INP_START] = 32778;
+    global.DEFAULT_KEY[3][global.INP_UP] = 87;
+    global.DEFAULT_KEY[3][global.INP_DOWN] = 83;
+    global.DEFAULT_KEY[3][global.INP_LEFT] = 65;
+    global.DEFAULT_KEY[3][global.INP_RIGHT] = 68;
+    global.DEFAULT_KEY[3][global.INP_FIRE1] = 69;
+    global.DEFAULT_KEY[3][global.INP_FIRE2] = 82;
+    global.DEFAULT_KEY[3][global.INP_START] = 33;
+    global.DEFAULT_JOY[3][global.INP_UP] = 32781;
+    global.DEFAULT_JOY[3][global.INP_DOWN] = 32782;
+    global.DEFAULT_JOY[3][global.INP_LEFT] = 32783;
+    global.DEFAULT_JOY[3][global.INP_RIGHT] = 32784;
+    global.DEFAULT_JOY[3][global.INP_FIRE1] = 32771;
+    global.DEFAULT_JOY[3][global.INP_FIRE2] = 32769;
+    global.DEFAULT_JOY[3][global.INP_START] = 32778;
+    global.DEFAULT_KEY[4][global.INP_UP] = 84;
+    global.DEFAULT_KEY[4][global.INP_DOWN] = 71;
+    global.DEFAULT_KEY[4][global.INP_LEFT] = 70;
+    global.DEFAULT_KEY[4][global.INP_RIGHT] = 72;
+    global.DEFAULT_KEY[4][global.INP_FIRE1] = 78;
+    global.DEFAULT_KEY[4][global.INP_FIRE2] = 66;
+    global.DEFAULT_KEY[4][global.INP_START] = 33;
+    global.DEFAULT_JOY[4][global.INP_UP] = 32781;
+    global.DEFAULT_JOY[4][global.INP_DOWN] = 32782;
+    global.DEFAULT_JOY[4][global.INP_LEFT] = 32783;
+    global.DEFAULT_JOY[4][global.INP_RIGHT] = 32784;
+    global.DEFAULT_JOY[4][global.INP_FIRE1] = 32771;
+    global.DEFAULT_JOY[4][global.INP_FIRE2] = 32769;
+    global.DEFAULT_JOY[4][global.INP_START] = 32778;
+    global.DEFAULT_KEY[5][global.INP_UP] = 219;
+    global.DEFAULT_KEY[5][global.INP_DOWN] = 222;
+    global.DEFAULT_KEY[5][global.INP_LEFT] = 186;
+    global.DEFAULT_KEY[5][global.INP_RIGHT] = 220;
+    global.DEFAULT_KEY[5][global.INP_FIRE1] = 190;
+    global.DEFAULT_KEY[5][global.INP_FIRE2] = 191;
+    global.DEFAULT_KEY[5][global.INP_START] = 33;
+    global.DEFAULT_JOY[5][global.INP_UP] = 32781;
+    global.DEFAULT_JOY[5][global.INP_DOWN] = 32782;
+    global.DEFAULT_JOY[5][global.INP_LEFT] = 32783;
+    global.DEFAULT_JOY[5][global.INP_RIGHT] = 32784;
+    global.DEFAULT_JOY[5][global.INP_FIRE1] = 32771;
+    global.DEFAULT_JOY[5][global.INP_FIRE2] = 32769;
+    global.DEFAULT_JOY[5][global.INP_START] = 32778;
+    for (p = 0; p <= 5; p++)
     {
         for (i = 0; i < global.NUM_INPUTS; i++)
         {
@@ -61,6 +117,10 @@
     }
     global.doubledUpButtons[0] = scrReadConfig("doubledUpButtonsP1", true);
     global.doubledUpButtons[1] = scrReadConfig("doubledUpButtonsP2", true);
+    global.doubledUpButtons[2] = scrReadConfig("doubledUpButtonsP3", true);
+    global.doubledUpButtons[3] = scrReadConfig("doubledUpButtonsP4", true);
+    global.doubledUpButtons[4] = scrReadConfig("doubledUpButtonsP5", true);
+    global.doubledUpButtons[5] = scrReadConfig("doubledUpButtonsP6", true);
     gamepad_set_axis_deadzone(0, 0);
     gamepad_set_axis_deadzone(1, 0);
     for (i = 0; i < 223; i++)
--- original/gml_GlobalScript_scrInputClear.gml
+++ patched/gml_GlobalScript_scrInputClear.gml
@@ -23,7 +23,7 @@
     pressStart = false;
     if (argument_count == 0 || argument[0] == true)
     {
-        for (p = 0; p <= 1; p++)
+        for (p = 0; p <= 5; p++)
         {
             joyTiltV[p] = 0;
             joyTiltH[p] = 0;
--- original/gml_GlobalScript_scrMenuNavigation.gml
+++ patched/gml_GlobalScript_scrMenuNavigation.gml
@@ -198,28 +198,28 @@
     {
         if (keyboard_check_pressed(vk_left) || keyboard_check_pressed(vk_right) || keyboard_check_pressed(vk_enter))
         {
-            if (keyboard_check_pressed(vk_left))
+            if (keyboard_check_pressed(vk_left) && itemIndex[menuSel] != 0)
             {
-                arrowLefttActive = arrowActiveMax;
+                itemIndex[menuSel] -= 1;
             }
-            if (keyboard_check_pressed(vk_right))
+            if (keyboard_check_pressed(vk_right) && itemIndex[menuSel] != 5)
             {
-                arrowRightActive = arrowActiveMax;
+                itemIndex[menuSel] += 1;
             }
-            itemIndex[menuSel] = !itemIndex[menuSel];
+            itemIndex[menuSel] = itemIndex[menuSel];
             return itemIndex[menuSel];
         }
         else if (pressLeft || pressRight || fire2pressed)
         {
-            if (pressLeft)
+            if (pressLeft && itemIndex[menuSel] != 0)
             {
-                arrowLefttActive = arrowActiveMax;
+                itemIndex[menuSel] -= 1;
             }
-            if (pressRight)
+            if (pressRight && itemIndex[menuSel] != 5)
             {
-                arrowRightActive = arrowActiveMax;
+                itemIndex[menuSel] += 1;
             }
-            itemIndex[menuSel] = !itemIndex[menuSel];
+            itemIndex[menuSel] = itemIndex[menuSel];
             return itemIndex[menuSel];
         }
     }
--- original/gml_GlobalScript_scrSaveConfig.gml
+++ patched/gml_GlobalScript_scrSaveConfig.gml
@@ -11,7 +11,7 @@
     scrWriteConfig("sfxToggle", global.AUDIO_SFX_TOGGLE);
     scrWriteConfig("bgmToggle", global.AUDIO_BGM_TOGGLE);
     scrWriteConfig("inputFocus", global.inputFocus);
-    for (p = 0; p <= 1; p++)
+    for (p = 0; p <= 5; p++)
     {
         for (i = 0; i < global.NUM_INPUTS; i++)
         {
@@ -24,6 +24,10 @@
     }
     scrWriteConfig("doubledUpButtonsP1", global.doubledUpButtons[0]);
     scrWriteConfig("doubledUpButtonsP2", global.doubledUpButtons[1]);
+    scrWriteConfig("doubledUpButtonsP3", global.doubledUpButtons[2]);
+    scrWriteConfig("doubledUpButtonsP4", global.doubledUpButtons[3]);
+    scrWriteConfig("doubledUpButtonsP5", global.doubledUpButtons[4]);
+    scrWriteConfig("doubledUpButtonsP6", global.doubledUpButtons[5]);
     scrWriteConfig("joyTypeP1", global.joyType[0]);
     scrWriteConfig("joyTypeP2", global.joyType[1]);
     scrWriteConfig("defaultLanguage", global.LANG_SAVE_ID[global.defaultLanguage]);
--- original/gml_Object_oPauseMenu_Other_12.gml
+++ patched/gml_Object_oPauseMenu_Other_12.gml
@@ -8,7 +8,7 @@
 if (substate == SUB_PRE_INIT)
 {
     currPlayer = 0;
-    for (p = 0; p <= 1; p++)
+    for (p = 0; p <= 5; p++)
     {
         for (i = 0; i < global.NUM_INPUTS; i++)
         {
@@ -141,7 +141,7 @@
             if (_isValid)
             {
                 scrSfxLibrary(soundSet[currentSoundSet]);
-                for (p = 0; p <= 1; p++)
+                for (p = 0; p <= 5; p++)
                 {
                     for (i = 0; i < global.NUM_INPUTS; i++)
                     {
@@ -155,7 +155,7 @@
                     }
                 }
                 keyMap[currPlayer][currInput] = currKey;
-                for (p = 0; p <= 1; p++)
+                for (p = 0; p <= 5; p++)
                 {
                     for (i = 0; i < global.NUM_INPUTS; i++)
                     {
@@ -180,14 +180,14 @@
 {
     if (stateCounter == 0)
     {
-        for (p = 0; p <= 1; p++)
+        for (p = 0; p <= 5; p++)
         {
             for (i = 0; i < global.NUM_INPUTS; i++)
             {
                 keyMap[p][i] = global.DEFAULT_KEY[p][i];
             }
         }
-        for (p = 0; p <= 1; p++)
+        for (p = 0; p <= 5; p++)
         {
             for (i = 0; i < global.NUM_INPUTS; i++)
             {
--- original/gml_Object_oPauseMenu_Other_13.gml
+++ patched/gml_Object_oPauseMenu_Other_13.gml
@@ -8,7 +8,7 @@
 if (substate == SUB_PRE_INIT)
 {
     currPlayer = 0;
-    for (p = 0; p <= 1; p++)
+    for (p = 0; p <= 5; p++)
     {
         for (i = 0; i < global.NUM_INPUTS; i++)
         {
@@ -193,7 +193,7 @@
                     }
                 }
                 joyMap[currPlayer][currInput] = currJoy;
-                for (p = 0; p <= 1; p++)
+                for (p = 0; p <= 5; p++)
                 {
                     for (i = 0; i < global.NUM_INPUTS; i++)
                     {
@@ -219,14 +219,14 @@
 {
     if (stateCounter == 0)
     {
-        for (p = 0; p <= 1; p++)
+        for (p = 0; p <= 5; p++)
         {
             for (i = 0; i < global.NUM_INPUTS; i++)
             {
                 joyMap[p][i] = global.DEFAULT_JOY[p][i];
             }
         }
-        for (p = 0; p <= 1; p++)
+        for (p = 0; p <= 5; p++)
         {
             for (i = 0; i < global.NUM_INPUTS; i++)
             {
@@ -235,6 +235,10 @@
         }
         global.doubledUpButtons[0] = true;
         global.doubledUpButtons[1] = true;
+        global.doubledUpButtons[2] = true;
+        global.doubledUpButtons[3] = true;
+        global.doubledUpButtons[4] = true;
+        global.doubledUpButtons[5] = true;
         global.joyType[0] = scrDetectJoyType(0);
         global.joyType[1] = scrDetectJoyType(1);
         stateCounter = 1;
