--- original/gml_GlobalScript_scrInitInput.gml
+++ patched/gml_GlobalScript_scrInitInput.gml
@@ -1,8 +1,10 @@
 function scrInitInput()
 {
     scrOpenConfig();
-    global.joySlot[0] = 0;
-    global.joySlot[1] = 1;
+    for (var p = 0; p <= 5; p++)
+    {
+        global.joySlot[p] = p;
+    }
     global.joyConnected = gamepad_is_connected(0);
     global.inputFocus = scrReadConfig("inputFocus", global.joyConnected + 1);
     if (!global.joyConnected)
@@ -48,7 +50,63 @@
     global.DEFAULT_JOY[1][global.INP_FIRE1] = 32771;
     global.DEFAULT_JOY[1][global.INP_FIRE2] = 32769;
     global.DEFAULT_JOY[1][global.INP_START] = 32778;
-    for (var p = 0; p <= 1; p++)
+    global.DEFAULT_KEY[2][global.INP_UP] = 73;
+    global.DEFAULT_KEY[2][global.INP_DOWN] = 75;
+    global.DEFAULT_KEY[2][global.INP_LEFT] = 74;
+    global.DEFAULT_KEY[2][global.INP_RIGHT] = 76;
+    global.DEFAULT_KEY[2][global.INP_FIRE1] = 85;
+    global.DEFAULT_KEY[2][global.INP_FIRE2] = 73;
+    global.DEFAULT_KEY[2][global.INP_START] = 33;
+    global.DEFAULT_JOY[2][global.INP_UP] = 32781;
+    global.DEFAULT_JOY[2][global.INP_DOWN] = 32782;
+    global.DEFAULT_JOY[2][global.INP_LEFT] = 32783;
+    global.DEFAULT_JOY[2][global.INP_RIGHT] = 32784;
+    global.DEFAULT_JOY[2][global.INP_FIRE1] = 32771;
+    global.DEFAULT_JOY[2][global.INP_FIRE2] = 32769;
+    global.DEFAULT_JOY[2][global.INP_START] = 32778;
+    global.DEFAULT_KEY[3][global.INP_UP] = 87;
+    global.DEFAULT_KEY[3][global.INP_DOWN] = 83;
+    global.DEFAULT_KEY[3][global.INP_LEFT] = 65;
+    global.DEFAULT_KEY[3][global.INP_RIGHT] = 68;
+    global.DEFAULT_KEY[3][global.INP_FIRE1] = 69;
+    global.DEFAULT_KEY[3][global.INP_FIRE2] = 82;
+    global.DEFAULT_KEY[3][global.INP_START] = 33;
+    global.DEFAULT_JOY[3][global.INP_UP] = 32781;
+    global.DEFAULT_JOY[3][global.INP_DOWN] = 32782;
+    global.DEFAULT_JOY[3][global.INP_LEFT] = 32783;
+    global.DEFAULT_JOY[3][global.INP_RIGHT] = 32784;
+    global.DEFAULT_JOY[3][global.INP_FIRE1] = 32771;
+    global.DEFAULT_JOY[3][global.INP_FIRE2] = 32769;
+    global.DEFAULT_JOY[3][global.INP_START] = 32778;
+    global.DEFAULT_KEY[4][global.INP_UP] = 84;
+    global.DEFAULT_KEY[4][global.INP_DOWN] = 71;
+    global.DEFAULT_KEY[4][global.INP_LEFT] = 70;
+    global.DEFAULT_KEY[4][global.INP_RIGHT] = 72;
+    global.DEFAULT_KEY[4][global.INP_FIRE1] = 78;
+    global.DEFAULT_KEY[4][global.INP_FIRE2] = 66;
+    global.DEFAULT_KEY[4][global.INP_START] = 33;
+    global.DEFAULT_JOY[4][global.INP_UP] = 32781;
+    global.DEFAULT_JOY[4][global.INP_DOWN] = 32782;
+    global.DEFAULT_JOY[4][global.INP_LEFT] = 32783;
+    global.DEFAULT_JOY[4][global.INP_RIGHT] = 32784;
+    global.DEFAULT_JOY[4][global.INP_FIRE1] = 32771;
+    global.DEFAULT_JOY[4][global.INP_FIRE2] = 32769;
+    global.DEFAULT_JOY[4][global.INP_START] = 32778;
+    global.DEFAULT_KEY[5][global.INP_UP] = 219;
+    global.DEFAULT_KEY[5][global.INP_DOWN] = 222;
+    global.DEFAULT_KEY[5][global.INP_LEFT] = 186;
+    global.DEFAULT_KEY[5][global.INP_RIGHT] = 220;
+    global.DEFAULT_KEY[5][global.INP_FIRE1] = 190;
+    global.DEFAULT_KEY[5][global.INP_FIRE2] = 191;
+    global.DEFAULT_KEY[5][global.INP_START] = 33;
+    global.DEFAULT_JOY[5][global.INP_UP] = 32781;
+    global.DEFAULT_JOY[5][global.INP_DOWN] = 32782;
+    global.DEFAULT_JOY[5][global.INP_LEFT] = 32783;
+    global.DEFAULT_JOY[5][global.INP_RIGHT] = 32784;
+    global.DEFAULT_JOY[5][global.INP_FIRE1] = 32771;
+    global.DEFAULT_JOY[5][global.INP_FIRE2] = 32769;
+    global.DEFAULT_JOY[5][global.INP_START] = 32778;
+    for (var p = 0; p <= 5; p++)
     {
         for (var i = 0; i < global.NUM_INPUTS; i++)
         {
@@ -59,8 +117,10 @@
             global.joyMap[p][i] = scrReadConfig("P" + string(p + 1) + "joy" + string(i), global.DEFAULT_JOY[p][i]);
         }
     }
-    global.doubledUpButtons[0] = scrReadConfig("doubledUpButtonsP1", true);
-    global.doubledUpButtons[1] = scrReadConfig("doubledUpButtonsP2", true);
+    for (var p = 0; p <= 5; p++)
+    {
+        global.doubledUpButtons[p] = scrReadConfig("doubledUpButtonsP" + string(p + 1), true);
+    }
     gamepad_set_axis_deadzone(0, 0);
     gamepad_set_axis_deadzone(1, 0);
     for (var i = 0; i < 223; i++)
@@ -304,8 +364,10 @@
     global.JOY_TYPE_SPRITE[n] = sJoyIconsGeneric;
     global.JOY_TYPE_NAME[n] = scrString("joy_type_generic");
     n++;
-    global.joyType[0] = scrReadConfig("joyTypeP1", scrDetectJoyType(0));
-    global.joyType[1] = scrReadConfig("joyTypeP2", scrDetectJoyType(1));
+    for (var p = 0; p <= 5; p++)
+    {
+        global.joyType[p] = scrReadConfig("joyTypeP" + string(p + 1), scrDetectJoyType(p));
+    }
     for (var i = 0; i <= 14; i++)
     {
         switch (i)
--- original/gml_GlobalScript_scrInputClear.gml
+++ patched/gml_GlobalScript_scrInputClear.gml
@@ -21,7 +21,7 @@
     pressStart = false;
     if (argument_count == 0 || argument[0] == true)
     {
-        for (var p = 0; p <= 1; p++)
+        for (var p = 0; p <= 5; p++)
         {
             joyTiltV[p] = 0;
             joyTiltH[p] = 0;
--- original/gml_GlobalScript_scrMenuNavigation.gml
+++ patched/gml_GlobalScript_scrMenuNavigation.gml
@@ -196,28 +196,26 @@
     {
         if (keyboard_check_pressed(vk_left) || keyboard_check_pressed(vk_right) || keyboard_check_pressed(vk_enter))
         {
-            if (keyboard_check_pressed(vk_left))
+            if (keyboard_check_pressed(vk_left) && itemIndex[menuSel] != 0)
             {
-                arrowLefttActive = arrowActiveMax;
+                itemIndex[menuSel] -= 1;
             }
-            if (keyboard_check_pressed(vk_right))
+            if (keyboard_check_pressed(vk_right) && itemIndex[menuSel] != 5)
             {
-                arrowRightActive = arrowActiveMax;
+                itemIndex[menuSel] += 1;
             }
-            itemIndex[menuSel] = !itemIndex[menuSel];
             return itemIndex[menuSel];
         }
         else if (pressLeft || pressRight || fire2pressed)
         {
-            if (pressLeft)
+            if (pressLeft && itemIndex[menuSel] != 0)
             {
-                arrowLefttActive = arrowActiveMax;
+                itemIndex[menuSel] -= 1;
             }
-            if (pressRight)
+            if (pressRight && itemIndex[menuSel] != 5)
             {
-                arrowRightActive = arrowActiveMax;
+                itemIndex[menuSel] += 1;
             }
-            itemIndex[menuSel] = !itemIndex[menuSel];
             return itemIndex[menuSel];
         }
     }
--- original/gml_GlobalScript_scrSaveConfig.gml
+++ patched/gml_GlobalScript_scrSaveConfig.gml
@@ -9,7 +9,7 @@
     scrWriteConfig("sfxToggle", global.AUDIO_SFX_TOGGLE);
     scrWriteConfig("bgmToggle", global.AUDIO_BGM_TOGGLE);
     scrWriteConfig("inputFocus", global.inputFocus);
-    for (var p = 0; p <= 1; p++)
+    for (var p = 0; p <= 5; p++)
     {
         for (var i = 0; i < global.NUM_INPUTS; i++)
         {
@@ -20,10 +20,11 @@
             scrWriteConfig("P" + string(p + 1) + "joy" + string(i), global.joyMap[p][i]);
         }
     }
-    scrWriteConfig("doubledUpButtonsP1", global.doubledUpButtons[0]);
-    scrWriteConfig("doubledUpButtonsP2", global.doubledUpButtons[1]);
-    scrWriteConfig("joyTypeP1", global.joyType[0]);
-    scrWriteConfig("joyTypeP2", global.joyType[1]);
+    for (var p = 0; p <= 5; p++)
+    {
+        scrWriteConfig("doubledUpButtonsP" + string(p + 1), global.doubledUpButtons[p]);
+        scrWriteConfig("joyTypeP" + string(p + 1), global.joyType[p]);
+    }
     scrWriteConfig("defaultLanguage", global.LANG_SAVE_ID[global.defaultLanguage]);
     scrCloseConfig();
 }
--- original/gml_Object_oPauseMenu_Other_12.gml
+++ patched/gml_Object_oPauseMenu_Other_12.gml
@@ -6,7 +6,7 @@
 if (substate == SUB_PRE_INIT)
 {
     currPlayer = 0;
-    for (var p = 0; p <= 1; p++)
+    for (var p = 0; p <= 5; p++)
     {
         for (var i = 0; i < global.NUM_INPUTS; i++)
         {
@@ -140,7 +140,7 @@
             if (_isValid)
             {
                 scrSfxLibrary(soundSet[currentSoundSet]);
-                for (var p = 0; p <= 1; p++)
+                for (var p = 0; p <= 5; p++)
                 {
                     for (var i = 0; i < global.NUM_INPUTS; i++)
                     {
@@ -154,7 +154,7 @@
                     }
                 }
                 keyMap[currPlayer][currInput] = currKey;
-                for (var p = 0; p <= 1; p++)
+                for (var p = 0; p <= 5; p++)
                 {
                     for (var i = 0; i < global.NUM_INPUTS; i++)
                     {
@@ -179,14 +179,14 @@
 {
     if (stateCounter == 0)
     {
-        for (var p = 0; p <= 1; p++)
+        for (var p = 0; p <= 5; p++)
         {
             for (var i = 0; i < global.NUM_INPUTS; i++)
             {
                 keyMap[p][i] = global.DEFAULT_KEY[p][i];
             }
         }
-        for (var p = 0; p <= 1; p++)
+        for (var p = 0; p <= 5; p++)
         {
             for (var i = 0; i < global.NUM_INPUTS; i++)
             {
--- original/gml_Object_oPauseMenu_Other_13.gml
+++ patched/gml_Object_oPauseMenu_Other_13.gml
@@ -7,7 +7,7 @@
 if (substate == SUB_PRE_INIT)
 {
     currPlayer = 0;
-    for (var p = 0; p <= 1; p++)
+    for (var p = 0; p <= 5; p++)
     {
         for (var i = 0; i < global.NUM_INPUTS; i++)
         {
@@ -31,8 +31,10 @@
         }
     }
     padCount = array_length(slotList);
-    gamepadConnect[0] = gamepad_is_connected(global.joySlot[0]);
-    gamepadConnect[1] = gamepad_is_connected(global.joySlot[1]);
+    for (var p = 0; p <= 5; p++)
+    {
+        gamepadConnect[p] = gamepad_is_connected(global.joySlot[p]);
+    }
     drawMenu = true;
     scrMenuCreate(scrString("menu_head_joypad_config"), menuSel);
     OP_PLAYER = scrMenuItem(TYPE_PLAYER, scrString("menu_item_player"), currPlayer);
@@ -211,7 +213,7 @@
                     }
                 }
                 joyMap[currPlayer][currInput] = currJoy;
-                for (var p = 0; p <= 1; p++)
+                for (var p = 0; p <= 5; p++)
                 {
                     for (var i = 0; i < global.NUM_INPUTS; i++)
                     {
@@ -237,24 +239,28 @@
 {
     if (stateCounter == 0)
     {
-        for (var p = 0; p <= 1; p++)
+        for (var p = 0; p <= 5; p++)
         {
             for (var i = 0; i < global.NUM_INPUTS; i++)
             {
                 joyMap[p][i] = global.DEFAULT_JOY[p][i];
             }
         }
-        for (var p = 0; p <= 1; p++)
+        for (var p = 0; p <= 5; p++)
         {
             for (var i = 0; i < global.NUM_INPUTS; i++)
             {
                 global.joyMap[p][i] = joyMap[p][i];
             }
         }
-        global.doubledUpButtons[0] = true;
-        global.doubledUpButtons[1] = true;
-        global.joyType[0] = scrDetectJoyType(0);
-        global.joyType[1] = scrDetectJoyType(1);
+        for (var p = 0; p <= 5; p++)
+        {
+            global.doubledUpButtons[p] = true;
+        }
+        for (var p = 0; p <= 5; p++)
+        {
+            global.joyType[p] = scrDetectJoyType(p);
+        }
         stateCounter = 1;
     }
     if (stateCounter == 1)
